src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,343 @@
+            }
+            streamLog[streamLog.size()-1].offset = std::get<0>(resp);
+            std::vector<seastar::future<std::tuple<Status, Payload> > > readFutures;
+            for (auto& log_entry: streamLog){
+                readFutures.push_back(_client.read(log_entry.name, 0, log_entry.offset));
+            }
+            return seastar::when_all_succeed(readFutures.begin(), readFutures.end());
+        })
+        .then([this] (std::vector<std::tuple<Status, Payload> >&& responses){
+            std::vector<Payload> newPayloads;
+            for (auto& response: responses){
+                auto& [status, resp] = response;
+                if (!status.is2xxOK()){
+                    throw k2::dto::PlogReadError("unable to read a plog");
+                }
+                newPayloads.push_back(std::move(resp));
+            }
+            return seastar::make_ready_future<std::vector<Payload> >(std::move(newPayloads));
+        });
+    });
+}
+
+
+seastar::future<> 
+LogStream::write(Payload payload){
+    return _write(std::move(payload), true);
+}
+
+seastar::future<> 
+LogStream::_write(Payload payload, bool writeToWAL){
+    if (!_logStreamCreate){
+        K2INFO("Log stream does not create");
+        throw k2::dto::LogStreamExistError("Log stream does not create");
+    }
+    std::pair<String, uint32_t>& targetPlogInfo = writeToWAL ? _walInfo : _metaInfo;
+
+    if (targetPlogInfo.second + payload.getSize() < PLOG_MAX_SIZE){
+        return _client.append(targetPlogInfo.first, targetPlogInfo.second, std::move(payload)).
+        then([this, writeToWAL] (auto&& response){
+            auto& [status, resp] = response;
+            if (!status.is2xxOK()) {
+                throw k2::dto::PlogAppendError("unable to append the plog");
+            }
+            // Check weather there is a sealed request that did not receive the response
+            if (_switched && writeToWAL){
+                _requestWaiters.emplace_back(seastar::promise<>());
+                // if there is a flying sealed request, we should not notify the client until we receive the response of that sealed request
+                return _requestWaiters.back().get_future().
+                then([] (){
+                    return seastar::make_ready_future<>();
+                });
+            }
+            else{
+                return seastar::make_ready_future<>();
+            }
+        });
+    }
+    else{
+        // switch to a new plog
+        return _switchPlog(std::move(payload), writeToWAL);
+    }
+}
+
+
+seastar::future<> 
+LogStream::_switchPlog(Payload payload, bool writeToWAL){
+    std::pair<String, uint32_t>& targetPlogInfo = writeToWAL ? _walInfo : _metaInfo;
+
+    String new_plogId;
+    String sealed_plogId = targetPlogInfo.first;
+    uint32_t sealed_offest = targetPlogInfo.second;
+    
+    // change to the next available plog id
+    if (!writeToWAL){

```

Comment:
Mingle WAL and meta data write into the same function is not a good idea due to coupling. No need to change it now, but we should naturally decouple them when we split wal and meta data from logstream.

Reply:
Listed a TODO here